### DISCUSSION TITLE
RDFLib V6 and V7

### DISCUSSION THREAD

**Discussion by aucampia** [RDFLib/rdflib#2395] [Development]
RDFLib has many open issues and known problems. My strategy for maintenance has been to try and fix as many of them without breaking the API, as breaking the API a little bit in each release is not really reasonable for our users, and if we take that strategy it seems we should rather abandon semantic versioning.

But there are some issues that are rather controversial to fix for V6. Many of them related to Datasets, but there are others also, most of these should show up in these searches:

- https://github.com/RDFLib/rdflib/labels/concept%3A%20RDF%20dataset
- https://github.com/RDFLib/rdflib/labels/id-as-cntxt
- https://github.com/RDFLib/rdflib/labels/backwards%20incompatible
- https://github.com/RDFLib/rdflib/labels/interface%20bug
- <https://github.com/RDFLib/rdflib/labels/7.0>
- https://github.com/RDFLib/rdflib/labels/breaking%20change

I was of the opinion that the majority of Dataset issues can be fixed with with the V6 API, but there seems to be disagreement on what the V6 API means, and they seem irreconcilable. Two key examples are:

- https://github.com/RDFLib/rdflib/issues/2393
- https://github.com/RDFLib/rdflib/issues/1804

These are really quite critical. If we can't handle Datasets correctly, we are not compliant with RDF 1.1 - and if we can't become compliant with RDF 1.1 because the expected behaviour is that we don't support RDF 1.1 - the only sane thing to do is to focus all energy on V7.

I think the right strategy for V7 should be to design a good API that fixes as many problems as we can with the current API and then implement it. I know @gjhiggins has started on some work towards redesign, but there has been little discussion on this and there is no published design.

I don't think trying to fix problems in V6 is productive any more, so I think the responsible thing is to abandon it and start with a new API. If anyone disagrees, please let me know. We should be realistic about V7 though, I think it will take at least a year to get there, if not more, as there are very few people working on it. And once it is released, most people who use RDFLib will have to rewrite their software. We could write a compatibility layer for V6 to use the new API, but I don't think there will be time for that.

CC: @RDFLib/core @RDFLib/core-reviewers

**Comment by aucampia** [comment]
~~I will merge this next week:~~

- ~~https://github.com/RDFLib/rdflib/pull/2108~~

~~The once that is merged, I will open a draft PR for the V7 API design. I will start with [Protocols](https://peps.python.org/pep-0544/) - we can then maybe proceed to abstract base classes from there. All of it will go into `rdflib.v7._provisional`.~~

~~I think we should look at the Jena and RDF4J APIs as some inspiration, possibly there are other APIs we should also look at. If anyone has a wish list for what they want to see in the V7 API let me know, I think it should be as minimal as possible. There is some chance that we can wrap re-implement some existing logic using the V7 concepts, so the right place to start is likely with:~~

- ~~Graph~~
- ~~Dataset~~
- ~~Serializer~~
- ~~Deserializer~~
- ~~Store~~

~~We should get as wide as possible review for the API, I will post it to various mailing lists.~~

~~Once we have an API we can start implementing abstract base classes where needed, and then proceed to write the code.~~

~~To ensure that we don't run into problems like [LSP-violations](https://github.com/RDFLib/rdflib/issues?q=is%3Aissue+is%3Aopen+LSP+label%3ALSP-violation) we should use strict typing for v7 code.~~

**Reply by nicholascar** [reply]
I very much support sorting out Datasets!

**Comment by aucampia** [comment]
~~The new API should also support RDFStar / RDF 1.2 as much as possible, and we should try and figure out how to support N3 in a way that retains type safety for RDF.~~

**Reply by nicholascar** [reply]
Don't forget this almost complete implementation of RDF-Star & SPARQL-Start for RDFLib: https://github.com/RDFLib/rdflib/pull/2115

**Reply by ajnelson-nist** [reply]
On type safety and RDF 1.1: here is one discussion spot where type safety, N3, and whether a triple's subject can be a `Literal` had previously come up: https://github.com/RDFLib/rdflib/pull/1853#issuecomment-1137723642

**Comment by aucampia** [comment]
I have reconsider this a bit. I think this is too ambitious and likely to fail. I think the right answer is just to not worry so much about breaking changes and just try do it as gradually as possible. The downside of this is that we will likely be releasing new major versions every time, but at least we can start fixing some of the interface bugs we have.

**Reply by ajnelson-nist** [reply]
Do you have a feel on what the first major version bump's leading Issue will be?  Handling Datasets correctly seems to be a goal necessitating a major version with several related Issues.

[Milestone 13](https://github.com/RDFLib/rdflib/milestone/13) seems to be a spot where next-major goals had been pooling.  Is it time to revive GitHub Milestones, or maybe a GitHub Project, to show what the current goal is, possibly primed with one of the Dataset issues you'd listed above?

**Reply by aucampia** [reply]
@ajnelson-nist I'm going to merge PRs that break interface as they are ready, and smaller PRs will be merged first. Right now we have no PRs ready though so it is hard to say. These two issues really bother me quite a bit and I can easily fix them:

- https://github.com/RDFLib/rdflib/issues/2393
- https://github.com/RDFLib/rdflib/issues/1804

I think @gjhiggins will also make some PRs soon.

The milestones are not really actively maintained, and I should really clean them up. The issue labels should also be cleaned up a bit, I don't think v7.0 label makes much sense. We should have a label indicating something is a breaking change, but then it goes in when it goes in.

Just as a reminder, there are no full-time maintainers. I only spend my personal time on RDFLib, I don't even use it for work.

The strategy going forward will basically be to just encourage PRs that do fix interface problems and make sure we fix them as well as possible, but try to fix singular interface issues in one PR instead of going for a drastic rewrite.

The consequence will be that for the next couple of releases we will likely have a major release every time, but our options are either that or stagnation. I don't think a clean slate API will work well.

I think possibly having some things in _rdlib.future in GitHub that represents an ideal API is not a bad idea, but that won't necessarily be in the release.

**Reply by aucampia** [reply]
I think it is important to note, changes that break the API should ideally be discussed up front in an issue to get some feedback before making a change. We won't reject PRs if there is no upfront discussion, but the PRs may need significant rework.

And the issues should be as singular as possible to make it easier to find consensus, and to make it easier to keep PR small.

**Reply by aucampia** [reply]
I made a PR to add this guidance to the RDFLib developers guide now: https://github.com/RDFLib/rdflib/pull/2402

**Reply by mielvds** [reply]
Not sure how to jump in here... but me and my organization (http://meemoo.be) use RDFlib actively, so we'll start spending some more time on contributing.

Now with respect to the versions and API: 
- is it realistic to preserve the (inaccurate) semantics of interfaces while introducing new implementations, and then deprecating the old ones one major version later? I.E. you would have a `Dataset` class living in a new namespace. If yes, I would go for that.
- regardless of the above, we would still need consensus on the semantics of the interface. Should we start with a consensus on expected behaviour and the unittests then? 
- I don't think that a sequence of major versions is an issue as long as we stick to semver
- I wouldn't refrain that much from making breaking changes. This happens all the time. As long as they are well documented and well versioned, the people maintaining the code have final say.

**Reply by aucampia** [reply]
> Not sure how to jump in here...

This is great, and thanks for the feedback, the more, the better. If you have time, do have a look at the PR that adds guidelines for breaking changes: https://github.com/RDFLib/rdflib/pull/2402 - there is also the matrix/gitter channel at https://matrix.to/#/#RDFLib_rdflib:gitter.im where I talk a lot though there is not always that much discussion.

> is it realistic to preserve the (inaccurate) semantics of interfaces while introducing new implementations, and then deprecating the old ones one major version later? I.E. you would have a Dataset class living in a new namespace. If yes, I would go for that.

I have thought about this (see details in https://github.com/RDFLib/rdflib/pull/2108 and https://github.com/RDFLib/rdflib/discussions/1841), but we really don't have the capacity for this I think, and it is really very ambitious.

I am, however, perfectly happy to accept PRs that add things in non-public namespaces, e.g:

- `rdflib._contrib`: For things that are useful but that we don't want to cover under semantic versioning guarantees.
- `rdflib._provisional`: For things that are likely to go into the public API without massive changes.
- `rdflib._future`: For aspirational things, this could maybe have a new API designed from scratch, but it will be very volatile.

But right now, I think the best option is to just fix the interface directly for most of the issues that exist, though I don't want to deter anyone from trying to get some consensus on a clean API built from scratch, though that will take capacity we don't really have, and I will prioritize actual fixes to the public API.

> regardless of the above, we would still need consensus on the semantics of the interface. Should we start with a consensus on expected behaviour and the unittests then?

Yes. I would say it is best to do all of it in one PR, but logical steps would be as you say, but I think it is also important to look at specific examples. <https://github.com/RDFLib/rdflib/pull/2406> is the first interface break that will be merged I think, as it is almost done, and simple. The clearest description of the problem it addresses is <https://github.com/RDFLib/rdflib/issues/2404>. 

It breaks the interface very subtly, to the point where maybe it can't be considered breaking it, but it fixes a serious issue which is that there is no way to read data into the default graph (or the closest thing we have to it) from sources that support named graphs. Here the expected behaviour is well established, I still started with a unit test, but it is going into the same PR as the fix.

**Reply by mielvds** [reply]
Those are some good points @aucampia . I've read through #2402 and I totally agree (I would like to see a note on deprecation though). Opposed of what I suggested earlier, I agree that fixing the existing interface with breaking changes is the most realistic and also less confusing. When bundled in a v7 release, the impact will be limited to one (major?) migration for users.

**Reply by aucampia** [reply]
> I would like to see a note on deprecation though

I will update the PR with some notes on this.

> When bundled in a v7 release, the impact will be limited to one (major?) migration for users.

We can only bundle PRs we get, but certainly, I hope we can bundle as many breaking changes as possible at once, though If I had to guess, I would say our next three releases will likely be v7, v8 and v9.

**Reply by aucampia** [reply]
@mielvds 

I added this to the PR regarding deprecation:

```rst
Deprecating functionality
~~~~~~~~~~~~~~~~~~~~~~~~~

To whatever extent possible, classes, functions, variables, or parameters that
will be removed should be marked for deprecation in documentation, and if
possible, should be changed to raise deprecation warnings if used.

There is however no hard requirement that something may only be removed after a
deprecation notice has been added, or only after a release was made with a
deprecation notice.

Consequently, functionality may be removed without it ever being marked as
deprecated.

.. _deprecation_rationale:

Rationale
^^^^^^^^^

Current resource limitations and the backlog of issues make it impractical to
first release or incorporate deprecation notices before making quality of life
changes.

RDFLib uses semantic versioning and provides type hints, and these are the
primary mechanisms for signalling breaking changes to our users.
```

I'm going to merge it this weekend if there is no further feedback. There is an insurmountable backlog, and we have to get cracking if RDFLib is to persist.

**Comment by gklyne** [comment]
Two thoughts (as someone who occasionally maintains an old codebase based originally on much older versions of RDFlib, and applies RDF updates infrequently and in large jumps):

1. I sense that many of the breaking changes relate to RDF 1.1 dataset support (I haven't looked at the individual issues).  It would be helpful to me if code that doesn't use datasets continues to work as before.
2. If there are breaking changes, make any breakages very noisy.  The worst thing in trying to keep an old codebase going is when changes cause subtle but silent errors.

**Reply by aucampia** [reply]
> I sense that many of the breaking changes relate to RDF 1.1 dataset support (I haven't looked at the individual issues). It would be helpful to me if code that doesn't use datasets continues to work as before.

It is hard to make guarantees, I think there is lots of work that is needed on the API, even the non-Dataset parts of it, but realistically I don't think it will be changed soon due to resource limitations. My priority is fixing concrete problems, and I will try my best to fix as much as possible with as few breaking changes as possible, but if we get good PRs that break the rest of the API for good reason then it will likely get accepted.

> If there are breaking changes, make any breakages very noisy. The worst thing in trying to keep an old codebase going is when changes cause subtle but silent errors.

Our mechanisms for signalling breaking changes are our version number, given we follow semantic versioning, our release notes and type hints. Of these, both semantic versioning and type hints provide mechanisms that facilitate automation. If you don't want breaking changes, the best option is to constrain your dependency on RDFLib to a specific major version. This is something you really should be doing anyway. Then, with type hints, if you do decide to upgrade to a new major version, the type hints, in conjunction with a static type checker like mypy, should signal at least some places where code will have to be updated with type errors.

**Reply by aucampia** [reply]
I should also mention that I'm trying to follow conventional commits, and we will possibly go further to ratify this, and use this to generate release notes. With this in place, the release notes should also clearly indicate breaking changes.

**Comment by ghost** [comment]
After #2406, I thought I'd revisit gromgull's PR #409 (a basic first step towards reworking Dataset) and investigate the ramifications.

I posted the resulting PR https://github.com/RDFLib/rdflib/pull/2438. Unhappily, some tests are failing.

If anyone wishes to review the original work, [gromgull's branch](https://github.com/gromgull/rdflib/tree/identifier-as-context) is forked from https://github.com/RDFLib/rdflib/commit/661a28938ffac1ed9898fd95dc48df2540e30de3  - for convenience: [here's the comparison](https://github.com/RDFLib/rdflib/compare/661a289..da6837e)
